File: internal/nano/task_recall.go
package nano

// task_recall.go is RETRIEVAL reframed as dense-supervised induction — the fix discovered after the
// canonical induction diagnostic showed the engine does induction perfectly, but only under DENSE
// supervision (the assoc/retrieval tasks supervised a single answer token and starved the circuit).
//
//   first half : [s1 v1 s2 v2 ... sK vK]          (K facts: distinct subjects, random values)
//   second half: [sπ1 vπ1 sπ2 vπ2 ... sπK vπK]    (the SAME subjects, SHUFFLED)
//   supervise  : at every second-half subject, predict its value (transform(value)) — K dense targets
//
// To answer at a second-half subject the model must find that subject among the K first-half facts
// (distractors) and copy/transform its value. That is genuine associative retrieval, now with the
// supervision density that lets the recall circuit form. With Transform off it is pure recall.

// RecallTask: M possible subjects, K facts/queries per sequence.
type RecallTask struct {
	M, K, L   int
	Transform bool // if true, target = (value+3) mod 10; else target = value (pure recall)
}

// NewRecallTask builds a recall task. L = 4K (K facts + K shuffled queries, each 2 tokens).
func NewRecallTask(m, k int, transform bool) RecallTask {
	return RecallTask{M: m, K: k, L: 4 * k, Transform: transform}
}

func (t RecallTask) subj(s int) int { return 10 + s }
func (t RecallTask) ans(v int) int {
	if t.Transform {
		return (v + 3) % 10
	}
	return v
}

func (t RecallTask) sample(rng *rng) (seq, targets []int) {
	// K distinct subjects with random values
	subs := make([]int, 0, t.K)
	seen := map[int]bool{}
	for len(subs) < t.K {
		s := int(rng.next() % uint64(t.M))
		if !seen[s] {
			seen[s] = true
			subs = append(subs, s)
		}
	}
	val := make(map[int]int, t.K)
	seq = make([]int, 0, t.L)
	for _, s := range subs {
		v := int(rng.next() % 10)
		val[s] = v
		seq = append(seq, t.subj(s), v)
	}
	// shuffled second half
	order := make([]int, len(subs))
	copy(order, subs)
	for i := len(order) - 1; i > 0; i-- {
		j := int(rng.next() % uint64(i+1))
		order[i], order[j] = order[j], order[i]
	}
	targets = make([]int, t.L)
	for i := range targets {
		targets[i] = -1
	}
	for _, s := range order {
		pos := len(seq) // position of this second-half subject token
		seq = append(seq, t.subj(s), val[s])
		targets[pos] = t.ans(val[s]) // predict the value (or its transform) AFTER the subject
	}
	return seq, targets
}

// Batch builds B sequences with K dense recall targets each.
func (t RecallTask) Batch(B int, seed int64) (idx, targets []int) {
	rng := newRNG(seed)
	idx = make([]int, B*t.L)
	targets = make([]int, B*t.L)
	for b := 0; b < B; b++ {
		seq, tg := t.sample(rng)
		copy(idx[b*t.L:], seq)
		copy(targets[b*t.L:], tg)
	}
	return
}

// Accuracy measures recall accuracy over the K second-half query positions of n fresh sequences.
func (t RecallTask) Accuracy(g *GPT, seed, n int) float64 {
	rng := newRNG(int64(seed))
	correct, total := 0, 0
	for s := 0; s < n; s++ {
		seq, tg := t.sample(rng)
		for pos := 0; pos < t.L; pos++ {
			if tg[pos] < 0 {
				continue
			}
			total++
			if g.PredictAt(seq, 1, t.L, pos) == tg[pos] {
				correct++
			}
		}
	}
	return float64(correct) / float64(total)
}

File: src/nano/task_recall.rs
//! RETRIEVAL reframed as dense-supervised induction. The first half lists K facts (subject value);
//! the second half repeats the SAME subjects shuffled, and at every second-half subject the model
//! must predict its value (or its transform) — K dense targets, the supervision density the recall
//! circuit needs.

use super::model::{Gpt, Rng};
use std::collections::{HashMap, HashSet};

/// RecallTask: M possible subjects, K facts/queries per sequence.
#[derive(Clone, Copy)]
pub struct RecallTask {
    pub m: usize,
    pub k: usize,
    pub l: usize,
    pub transform: bool, // if true, target = (value+3) mod 10; else target = value (pure recall)
}

impl RecallTask {
    /// NewRecallTask builds a recall task. L = 4K (K facts + K shuffled queries, each 2 tokens).
    pub fn new(m: usize, k: usize, transform: bool) -> RecallTask {
        RecallTask {
            m,
            k,
            l: 4 * k,
            transform,
        }
    }

    fn subj(&self, s: usize) -> usize {
        10 + s
    }
    fn ans(&self, v: usize) -> usize {
        if self.transform {
            (v + 3) % 10
        } else {
            v
        }
    }

    fn sample(&self, rng: &mut Rng) -> (Vec<usize>, Vec<i32>) {
        // K distinct subjects with random values
        let mut subs: Vec<usize> = Vec::with_capacity(self.k);
        let mut seen: HashSet<usize> = HashSet::new();
        while subs.len() < self.k {
            let s = (rng.next() % self.m as u64) as usize;
            if !seen.contains(&s) {
                seen.insert(s);
                subs.push(s);
            }
        }
        let mut val: HashMap<usize, usize> = HashMap::with_capacity(self.k);
        let mut seq: Vec<usize> = Vec::with_capacity(self.l);
        for &s in &subs {
            let v = (rng.next() % 10) as usize;
            val.insert(s, v);
            seq.push(self.subj(s));
            seq.push(v);
        }
        // shuffled second half
        let mut order = subs.clone();
        let mut i = order.len() - 1;
        while i > 0 {
            let j = (rng.next() % (i as u64 + 1)) as usize;
            order.swap(i, j);
            i -= 1;
        }
        let mut targets = vec![-1i32; self.l];
        for &s in &order {
            let pos = seq.len(); // position of this second-half subject token
            seq.push(self.subj(s));
            seq.push(val[&s]);
            targets[pos] = self.ans(val[&s]) as i32; // predict the value (or transform) AFTER the subject
        }
        (seq, targets)
    }

    /// Batch builds B sequences with K dense recall targets each.
    pub fn batch(&self, b: usize, seed: i64) -> (Vec<usize>, Vec<i32>) {
        let mut rng = Rng::new(seed);
        let mut idx = vec![0usize; b * self.l];
        let mut targets = vec![0i32; b * self.l];
        for bb in 0..b {
            let (seq, tg) = self.sample(&mut rng);
            idx[bb * self.l..bb * self.l + self.l].copy_from_slice(&seq);
            targets[bb * self.l..bb * self.l + self.l].copy_from_slice(&tg);
        }
        (idx, targets)
    }

    /// Accuracy measures recall accuracy over the K second-half query positions of n fresh sequences.
    pub fn accuracy(&self, g: &Gpt, seed: i64, n: usize) -> f64 {
        let mut rng = Rng::new(seed);
        let mut correct = 0;
        let mut total = 0;
        for _ in 0..n {
            let (seq, tg) = self.sample(&mut rng);
            for pos in 0..self.l {
                if tg[pos] < 0 {
                    continue;
                }
                total += 1;
                if g.predict_at(&seq, 1, self.l, pos) == tg[pos] as usize {
                    correct += 1;
                }
            }
        }
        correct as f64 / total as f64
    }
}
